Fixes #27853: Connect ForemanModal to Redux store

jeremylenz opened the pull request:
* Connect `ForemanModal` to Redux store
* Add `useForemanModal` hook for reusing stateful logic in any `ForemanModal` component
* Modal state is now controlled by Redux store
* Uses `react-redux` v7.1 which includes `useSelector` and allows use of Redux without `connect`, `mapStateToProps` or `mapDispatchToProps`

jeremylenz committed:
Fixes #27853: Connect ForemanModal to Redux store

For unsquashed commits, see jeremylenz/27853-modals-redux-unsquashed

Connect ForemanModal React component to Redux store.
Add useForemanModal hook
  for reusing stateful logic in any ForemanModal component.
Modal state is now controlled by id with Redux actions.
Uses react-redux v7.1,
  which includes useSelector hook
  and allows use of Redux without connect,
  mapStateToProps or mapDispatchToProps.

jeremylenz commented:
Leaving [WIP] until #7039 is merged
Updated 6 Nov 2019 - ready for review

jeremylenz commented:
test failures unrelated

jeremylenz commented:
Rebased after ForemanModal #7039 merged

xprazak2 commented:
Looking good, can't wait to use this.

The snapshot changes for bookmarks and forms are caused by newer react-redux version, right?

packaging: https://github.com/theforeman/foreman-packaging/pull/4302

jeremylenz commented:
> Looking good, can't wait to use this.
> 
> The snapshot changes for bookmarks and forms are caused by newer react-redux version, right?
> 
> packaging: [theforeman/foreman-packaging#4302](https://github.com/theforeman/foreman-packaging/pull/4302)

@xprazak2 Yes I believe so

jeremylenz commented:
Snapshots failing on Travis due to update of a patternfly-react component.

xprazak2 commented:
Needs https://github.com/theforeman/foreman/pull/6981, which is currently waiting on packaging, so setting as blocked by other PR.

amirfefer commented on webpack/assets/javascripts/react_app/components/FactCharts/__test__/FactChart.test.js:
```diff
@@ -22,12 +22,4 @@ describe('factCharts', () => {
     expect(toJson(wrapper)).toMatchSnapshot();
   });
-
-  it('should open modal after click', () => {
```
no test?

jeremylenz replied:
@amirfefer I'm not sure why this was deleted; I will add it back

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.js:
```diff
@@ -1,17 +1,24 @@
+  const openState = useSelector(state => selectModalStateById(state, id));
+  const isOpen = false || (openState && openState.open);
+  const dispatch = useDispatch();
+  const onClose = () => dispatch({ type: SET_MODAL_CLOSED, payload: { id } });
```
With redux hooks we lost our convention for separation between redux and pure react.
I'd prefer such separation because we can gain better unit testing and better workflow, could we make such a separation with hooks?

jeremylenz replied:
@amirfefer Thanks for bringing up the concern.  I hadn't really thought about it much, but after poking around it seems a lot of people in the JS community are talking about it.

Scroll down to the comments below this [Redux hooks article](https://dev.to/gsto/new-redux-hooks-a-before-and-after-comparison-are-they-better-loj) and there's some really good discussion, including someone who has the same concerns as you.  (I really like [this comment](https://dev.to/sschottler/comment/bmkn) and also want to call out [Dan Abramov's 2019 update](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0) to his 2015 article.

The gist seems to be that Hooks allow better separation of stateful logic and presentational logic (JSX), but on the other hand require closer coupling of other things.

Also today I checked out [React Testing Library](https://testing-library.com/docs/react-testing-library/intro), which is a replacement for Enzyme with a very different philosophy.  They're saying that instead of testing instances of React components, it makes more sense to look at the DOM elements that the components output, and then test by looking for text etc. on the screen, like a human would.  (Incidentally, this is also supposed to work better with Hooks than Enzyme does.)

As far as testing in Foreman, for now I'm proposing a solution with Enzyme in the comment below.

amirfefer replied:
Thanks for bringing this up! I think hooks are a great feature but still without best practices yet. I agree we need to change our conventions due to the evolving process, and the enzyme alternative looks totally interesting. 

However, IMHO React is going fast, and breaking our conventions over and over makes the code difficult to newcomers and increase our technical debt.
I do think we should evolve together with react, but changing our conventions should be a part of dedicated design, after a deeper understanding of the new Hooks and its best practices and  with a wide consensus (discourse thread would be a great start)

Until then I think something like [this](https://github.com/theforeman/foreman/pull/7105#discussion_r349004392) can work :)

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.test.js:
```diff
@@ -1,7 +1,29 @@
+  },
+};
+
+const ConnectedForemanModal = props => (
```
This should be a unit test, but because combining redux hooks it becomes more like an integration test.
another reason for a react/redux separation

jeremylenz replied:
(Continuing from above)  So, how to write an Enzyme test for a component with that uses Hooks and react-redux?

There seems to be two options:

1. Mock the redux store and use Enzyme `mount`, making it more of an integration test.  This is what I had originally.
2. Mock `react-redux` including any Hooks used in the component.  That way, it allows you to use `shallow` which is closer to a unit test of the component, and it doesn't care about the Redux implementation.  This is referenced in the 2nd answer to [this StackOverflow question](https://stackoverflow.com/questions/56827300/how-to-test-a-component-using-react-redux-hooks).  After coming across this, I really like this solution.

I added a new test file, `ForemanModalNoRedux.test.js` that uses this implementation.  It does the following:
* Mocks `useSelector` with a fake store state
* Mocks `useDispatch` with a fake `dispatch` function
* All Redux-connected values are mocked, so we are just testing the component itself
* Use `testComponentSnapshotWithFixtures` for all tests, meaning `shallow` rendering is used and thus is a unit test
* Move the two tests I was using `mount` for into the fixtures and test them without deep rendering

In this way, I think we can keep the testing of Hooks components separate from testing Redux.  If this is acceptable, I will make this file the new `ForemanModal.test.js` and eliminate the old one.

@sharvit @waldenraines @laviro @MariaAga @johnpmitsch @jturel   anyone else, would also love to hear your thoughts!

jeremylenz replied:
@amirfefer bumping the above comment :)

sharvit replied:
IMO we can continue using the current approach.

- unit-testing for the `Component.js`
Where we import the `Component.js` file, using `shallow` and `snapshots` to make sure the component gets rendered correctly with each set of props.

- integration testing for the `index.js`
Where we import the `index.js` file, using `mount` to test that all the different parts are working correctly within the redux lifecycle.
`InitialState` -> `ComponentRendering` -> `EventTriggering` -> `ActionCalled` -> `StateUpdated` -> `ComponentRendering`

TBH I don't like the approach of importing the `index.js`, mocking everything inside so I can write a unit-test.
The unit is the component and I can import it directly from the `Component.js` file without the need to use the `index.js` file that needs many other moving parts to work together.

jeremylenz replied:
> TBH I don't like the approach of importing the index.js, mocking everything inside so I can write a unit-test.

@sharvit Ok, I'm going to try redoing the tests again so that
* unit tests (`shallow` and your unit test helpers) apply to our (newly presentational) `ForemanModal.js`, and
* integration tests (`mount`) apply to our container component `index.js`.

jeremylenz replied:
@sharvit @amirfefer Refactored the tests as stated above, ready for another look

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.test.js:
```diff
@@ -1,7 +1,29 @@
+};
+
+const ConnectedForemanModal = props => (
+  <Provider store={mockStore(state)}>
```
you can use integration test helper for creating the store for you

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalReducer.js:
```diff
@@ -0,0 +1,56 @@
+            `SET_MODAL_OPEN error: Modal with id '${action.payload.id}' does not exist`
+          )
+        );
+      }
```
how about moving this error throwing to the action?

jeremylenz replied:
@amirfefer Good idea, will update

jeremylenz replied:
[update: see comment](https://github.com/theforeman/foreman/pull/7105#discussion_r348769976)

amirfefer replied:
Any thoughts about error handling via redux? which means the action creator will  throw an error instead, and as a result it dispatches an action that changes the state respectively to error state

jeremylenz replied:
@amirfefer After moving the error handling to Actions instead of Reducers, I'm still thinking Actions isn't the right place for it either.  After making the code change and thinking about it for a bit, I actually think middleware would be the correct place.  One totally legitimate job for middleware is to filter actions and conditionally dispatch them.  (And you'd have access to the current store state easily as well.)

But anyway, I put the error handling in action creators for now since that feels less incorrect than reducers.  Would love to get thoughts from @sharvit also.

sharvit replied:
Errors in javascript is a huge topic :)

When an Error gets `throw`n, the script will stop executing unless you `catch` it. And, it does not matter if you `throw` an Error manually, it is `throw`n by a library, or by the runtime itself (node or the browser).
Therefore, not catching errors is a bad practice since it might/will break the application in an unpredictable way.

IMO there are 2 types of errors
1. **Predicted errors** - The **user** did something wrong and the UI should respond and tell the **user** about it.
A form validation error is an example.

2. **Unpredicted errors** - The **developer** did something wrong and it can lead to unpredictable behavior.
We cannot respond to those errors but we do want to tell the **developer** about what went wrong using logs and it can be nice to keep to UI functioning :)
An example of it is the current error, the **developer** tries to open a model which is not exists.

**IMO it is a good practice having a middleware catching unpredicted errors caused by actions and log them so the developer will have clue about what went wrong.**

BTW, doesn't the `redux-logger` middleware we use in `development` shouldn't catch and log errors in actions by default?
Does it make sense to log those **developer** related errors in production?

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalReducer.js:
```diff
@@ -0,0 +1,56 @@
+      return state.merge(newState);
+    }
+    case ADD_MODAL: {
+      if (state[action.payload.id]) return state; // if it already exists, don't change its state
```
can we filter this in the action? This returns the same state but all the middlewares were activated for nothing

jeremylenz replied:
@amirfefer true, will update!

jeremylenz replied:
Update: I realized that in order to do this, we will need access to the store state from within the action creator.  This would require one of the following:

1. In `ForemanModalActions.js`, import the Redux store directly and use `store.getState()`
2. Make all of the ForemanModal actions into thunks and use the `getState` provided by `redux-thunk`
3. Pass the store state into the action creator as a function parameter.

I don't like any of these.  On the other hand, going through our middleware shouldn't really cause any noticeable performance issue because it would be filtered out by all the middleware logic and passed straight through to the reducer.

amirfefer replied:
while redux-thunk's `getState` is simple and has a constant cost, middlewares stack tends to grow, plus I'd prefer having cleaner reducers by moving logic (i.e side effects, filtering, error handling...) to actions.

jeremylenz replied:
@amirfefer I agree that of the 3 options, `getState` seems to be the lesser of the evils.  Stay tuned for update

amirfefer replied:
any update? :)

jeremylenz replied:
@amirfefer The action creator now handles the error messages, but I think I left that line in so that you can't set a modal open or closed with `ADD_MODAL`.  (You could theoretically still dispatch the plain action without using the action creator.)

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalReducer.js:
```diff
@@ -0,0 +1,56 @@
+          __(
+            `SET_MODAL_CLOSED error: Modal with id '${action.payload.id}' does not exist`
+          )
+        );
```
ditto

jeremylenz replied:
[see comment](https://github.com/theforeman/foreman/pull/7105#discussion_r348769976)

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/index.js:
```diff
@@ -1,3 +1,6 @@
 import ForemanModal from './ForemanModal';
+import reducer from './ForemanModalReducer';
+
```
how about creating all the hooks here, and pass it to the component as props? so we can keep our separation convention

jeremylenz replied:
@amirfefer Hooks must be used only inside a function component or in a custom Hook.  (But see the other comments for more discussion :yum: )

amirfefer replied:
How about this approach:

```js
// component/index.js

import { useSelector, useDispatch } from 'redux-react'
import Component from './Component'
import { selectAllData, selectSomeOtherData } from './ComponentSelectors'
import { hide } from './ComponentsActions'

const dispatch = useDispatch();

const ConnectedComponent = () =>
<Component
  allRecords=useSelector(selectAllData)
  otherData=useSelector(selectSomeOtherData)
  hide=dispatch(hide);
/>
```
  
With this we still gain a great separation between the markup and the state, we use hooks and keep our conventions. It probably causes extra code writing, but such a separation I  believe makes the code cleaner and readable.

jeremylenz replied:
```js
const dispatch = useDispatch();
```
@amirfefer This is a use of a hook outside a [React function](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-from-react-functions), so unfortunately it won't work.  You'd have to move the `useDispatch()` to within `ConnectedComponent`.

amirfefer reviewed (changes requested):
Thanks @jeremylenz ! looks great and works well (storybook tested)
My concern is about breaking the redux/react separation, do you think we can keep that convention with hooks?

jeremylenz commented:
Also updated the selector tests which were broken (the snapshots were showing undefined values; fixed now)

jeremylenz commented:
New changes yesterday and today:

1. Lift logic out of `ForemanModal.js` and into `index.js`
2. Move error handling from reducers to actions

I kept the commits separate, so you can use the Github feature when viewing changes to view changes from just a single commit.

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.js:
```diff
@@ -1,17 +1,16 @@
-    props.children
+    children
   );
+  const isOpen = false || (openState && openState.open);
```
`openState`'s  value is false by default, do we still need an `isOpen` variable?, how about  pass it as prop, something like this:
```js
const const { id, title, onClose, openState: { isOpen }  , children, ...propsToPassDown } = props;

const context = {
    isOpen,
    onClose,
}

```

jeremylenz replied:
true, I will get rid of the `isOpen` variable

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalNoRedux.test.js:
```diff
@@ -0,0 +1,84 @@
+import React from 'react';
```
Instead of `NoRedux` suffix we have a convention for tests files:
`Component.test.js` for a unit test (without redux)
`integration.test.js` for a react+redux test
`ComponentActions.test.js` for an actions test
`ComponentReducer.test.js` for  a reducer test

jeremylenz replied:
@amirfefer maybe you missed the [comment here](https://github.com/theforeman/foreman/pull/7105#discussion_r348710831).  The `ForemanModalNoRedux.test.js` is meant to replace `ForemanModal.test.js`. If you find it an acceptable approach I plan to rename it `ForemanModal.test.js`.

amirfefer replied:
@sharvit could you please look at this change of the test structure?

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalNoRedux.test.js:
```diff
@@ -0,0 +1,84 @@
+import { testComponentSnapshotsWithFixtures } from '../../common/testHelpers';
+
+jest.mock('react-redux', () => ({
+  useSelector: jest.fn(fn =>
```
can we move the hooks mocking mechanism to `testsHelpers`? to make it easier for reusing?

jeremylenz replied:
@amirfefer It would be nice but I don't think so, because `jest.mock` only works for the file it's called in.

jeremylenz replied:
Also the specific hooks that you need to mock may be different for each file.

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalReducer.js:
```diff
@@ -0,0 +1,41 @@
+      newState[action.payload.id] = {
+        open: action.payload.open || false,
+      };
+      return state.merge(newState);
```
you can use `setIn` instead of creating a new object:

```js
state.setIn([action.payload.id, 'open'],  action.payload.open || false);
```

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalReducer.js:
```diff
@@ -0,0 +1,41 @@
+  switch (action.type) {
+    case SET_MODAL_OPEN: {
+      const newState = {};
+      newState[action.payload.id] = {
```
ditto

amirfefer commented on webpack/assets/javascripts/react_app/components/ForemanModal/index.js:
```diff
@@ -1,3 +1,39 @@
+      openState={openState}
+      onClose={onClose}
+    />
+  );
```
looks great :+1:

amirfefer reviewed (dismissed):
Thanks @jeremylenz !  Looks promising :+1: 
I left a few inline comments

MariaAga commented:
I find it a bit confusing that the consumer can provide a title to the modal but it would be only used if the consumer uses `<ForemanModal.Header />`, is there a clearer way to do this?

jeremylenz commented:
@MariaAga you're right, it was a confusing API.  

First I thought, I'll just make it render a default `<ForemanModal.Header />` if you don't supply one.  That way the title will always be used.

But I also have to have a way for the ForemanModal not to have a header, because I want my component to be as flexible as possible.  So, I came up with this solution:

* The `title` prop remains a required PropType on `ForemanModal`
* If you don't supply a `<ForemanModal.Header>` as a child of `<ForemanModal>`, ForemanModal will render a default `<ForemanModal.Header>` with the `title`
* If you supply a `<ForemanModal.Header>`, ForemanModal will render your markup and use the `title`
* If you want to have a ForemanModal without a header, you can pass a `noHeader` prop to ForemanModal.  This will force the header not to be rendered.

Also updated storybook and tests.  Let me know what you think.

MariaAga commented:
@jeremylenz Thanks for the update, could you explain why `title` is required instead of making it optional and removing the `noHeader` props?

jeremylenz commented:
@MariaAga I guess I was just thinking "why wouldn't you want your modal to have a title?"  And also, why would you want a modal with a header but no title?

MariaAga commented:
@jeremylenz I think the default option should be minimal and if you want to add a header and a title you can do so
Also this is no longer blocked by another PR as https://github.com/theforeman/foreman/pull/7178 was merged

jeremylenz commented:
@MariaAga hmm, yes that may be simpler!  So how about:

* If you provide a `title` prop and no `<ForemanModal.Header>` child, the default header will be rendered with title
* If you provide a header child with no `title` prop, your own markup will be used for the header (and you can choose whether to have a title, etc. in the header)
* If you provide neither a header child nor a `title` prop, the modal won't have a header. 
* The `noHeader` prop won't be needed in any case.

does this sound good to you?

MariaAga commented:
@jeremylenz makes sense to me!

jeremylenz commented:
Also, I'm rethinking the API for the `useForemanModal` hook.  Currently it returns an array
```js
[modalOpen, toggleModal]
```
where `toggleModal()` is a function to close the modal if it's open, and vice versa.  But I'm thinking maybe it's more explicit and less error-prone if it returns an object like
```js
{
  modalOpen: [bool],
  setModalOpen: [func],
  setModalClosed: [func],
}
```
and has explicit functions for opening and closing, instead of deciding based on current state.  Thoughts?

MariaAga commented:
@jeremylenz This looks great and more readable!

jeremylenz commented:
Latest changes:

* Refactored tests as per [this comment](https://github.com/theforeman/foreman/pull/7105#discussion_r353279533)
* Refactored `useForemanModal` as per [this comment](https://github.com/theforeman/foreman/pull/7105#issuecomment-561236575)
* Updated Storybook with Redux and Hooks stories

sharvit commented on package.json:
```diff
@@ -20,7 +20,7 @@
   },
   "dependencies": {
-    "@theforeman/vendor": "^1.7.0",
+    "@theforeman/vendor": "^2.15.7",
```
Please rebase, foreman already uses ^3

jeremylenz replied:
Rebased

sharvit commented on webpack/assets/javascripts/react_app/components/FactCharts/__test__/FactChart.test.js:
```diff
@@ -22,12 +22,12 @@ describe('factCharts', () => {
-  it('should open modal after click', () => {
-    const props = { onClick: jest.fn(), data: { id: 1 } };
-    const wrapper = shallow(<FactChart store={Store} {...props} />);
+it('should open modal after click', () => {
```
Does it actually test something?

It seems to me that because it uses `shallow`, enzyme will just call the `onClick` prop without using the inner logic when you simulate `click`.

sharvit replied:
Actually this test fails for me.

jeremylenz replied:
@sharvit This was just re-adding a test that got deleted somehow.  I didn't mean to make any changes to `FactChart`.   let's see what happens after rebase..

jeremylenz replied:
@sharvit As I suspected, this test isn't present in `develop`.  Removing.

sharvit commented on webpack/assets/javascripts/react_app/components/ForemanModal/index.js:
```diff
@@ -1,3 +1,43 @@
+  const { id, title } = props;
+  const openState = useSelector(state => selectModalStateById(state, id));
+  const dispatch = useDispatch();
+  const onClose = () => dispatch({ type: SET_MODAL_CLOSED, payload: { id } });
```
Wouldn't it make more sense using the `setModalClosed` action here?

jeremylenz replied:
@sharvit For me it's about 50/50.  I really like the explicitness of being able to see the plain action when you read the code.  But action creators are also convenient and seems to be what we're used to in Foreman, so I'll change it

sharvit commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.test.js:
```diff
@@ -1,42 +1,60 @@
     testComponentSnapshotsWithFixtures(ForemanModal, fixtures);
   });
+  describe('PropTypes', () => {
+    it('requires an id prop', () => {
```
IMO there is no reason to test it, we can rely on `prop-types` that they are working.

jeremylenz replied:
@sharvit In general I totally agree that PropTypes don't need their own tests.  In this case I just thought that since `id` is such an important part of `ForemanModal`, and defines one of the essential features of the implementation, that it would be good to state that in test form.  I'm fine deleting the test if you still think it's unnecessary, but just wanted to explain my reason :)

jeremylenz replied:
@sharvit After adding the integration test I changed my mind.  Removing.

sharvit commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.stories.js:
```diff
@@ -1,152 +1,231 @@
+  .add('With default header & footer', () =>
+    // using createElement here so that hooks work in stories
+    React.createElement(() => {
+      const { setModalOpen } = useForemanModal({ id: 'default' });
```
I'm wondering why do we need the `useForemanModal`, IMO it becomes too complicated to use it.
ATM I need to use `redux` and `useForemanModal` together and it feels like they give me the same functionality (open, close, isOpen). 

It can be implemented differently with `redux` only.

1. on-mount - call the `addModalAction` with the `id`
2. use **only** `redux` actions to close and open
3. use **only** `redux` selectors to get your modal `isOpen` state

It will make my life as a consumer easier, to use the `ForemanModal` I will only need to:
```js
const MyExampleModalUsage = ({
  isMyExampleModalOpen,
  openMyExampleModal,
  closeMyExampleModal,
}) => (
  <div>
    <ForemanModal id="myExampleModal">...</ForemanModal>
    {!isMyExampleModalOpen && (
      <button onClick={openMyExampleModal}>open</button>
    )}
  </div>
);

const MyExampleModalUsageConnected = () => {
  const dispatch = useDispatch();
  return (
    <MyExampleModalUsage
      isMyExampleModalOpen={useSelector(selectIsModalOpen('myExampleModal'))}
      openMyExampleModal={useDispatch(() => setModalOpen('myExampleModal'))}
      closeMyExampleModal={useDispatch(() => setModalClose('myExampleModal'))}
    />
  );
};
```

jeremylenz replied:
@sharvit `useForemanModal` is totally optional.  It's totally valid to choose just to use plain Redux actions, or action creators, or the Hook.  I think any of those three approaches are fine.  (Also, the hook only works if your component is written using Hooks.  So it naturally can't be the only option.)

sharvit replied:
If we will allow consumers to achieve the same, using 2 different ways, it will be confusing for consumers about what they should choose.
It also means we should maintain 2 different interfaces for the long term.

jeremylenz replied:
@sharvit If we decide not to provide a Hook just because Hooks might be confusing, then we shouldn't allow the use of React hooks at all in Foreman.

Rather, I think Hooks are the future, and for a new learner classes are actually the more confusing option.  It's just because we don't have an "empty cup" and are used to classes that it's harder for us to see this as more experienced React devs.

As far as the long term, we should expect to write more and more with hooks and less with classes, and during the transition period we will have to live with it.  Just like we live with our Angular and jQuery today. ;) But I think it's worth it!

The developers of React do not deprecate class-based components, and they also offer hooks.  The developers of Redux have not deprecated `connect`, `mapStateToProps`, etc.  Just as they offer both options, I think it's fine for us to offer parallel options.

jeremylenz replied:
@sharvit If you use `useForemanModal` then you don't need to use Redux explicitly at all-- the hook uses it for you behind the scenes.  It's reusing the stateful logic (the ForemanModal-related selectors and dispatches) that you otherwise would have to rewrite every time you use ForemanModal.

However, there's always an argument for being more explicit.  So if you want to use `useSelector` and `useDispatch` every time, and not wrap it in `useForemanModal` that's just fine.

sharvit commented on webpack/assets/javascripts/react_app/components/ForemanModal/ConnectedForemanModal.test.js:
```diff
@@ -0,0 +1,118 @@
+const modalBody = <div>This is the body</div>;
+
+describe('ForemanModal with Redux & hooks', () => {
+  describe('rendering', () => {
```
It doesn't feel like an integration test.

Instead of testing your component works together with your reducers, actions selectors. It render with modal with a given state like you already do in your unit testing.

IMO it should use the `IntegrationTestHelper` to make sure all the different parts work together as a bigger unit.

Just sketching around, but I expected something like:
https://gist.github.com/sharvit/b7d193c033d9c5fe6371eb4efb27a467

jeremylenz replied:
yeah it's more like a fat unit test, right?  Thanks for the idea, I will play around with this..

jeremylenz replied:
@sharvit Updated the tests as per your suggestion, thanks :)

sharvit reviewed (changes requested):
Thanks @jeremylenz :+1: 

Where it comes to folder structure and conventions:
1. The test files should live under `__tests__` folder
2. The `subcomponents` folder should rename to `components`

Thanks!!

jeremylenz commented:
@sharvit  About the `subcomponents` folder name, as far as I'm aware this is the first component in Foreman that is a "compound component"-- meaning it has `<ForemanModal>` as the main component and also `<ForemanModal.Header>` and `<ForemanModal.Footer>` subcomponents which are parts of the main component.  It's not the same convention as having a Page which contains several components.   In light of this it makes sense to me to call the folder `subcomponents` and not just `components`.

jeremylenz commented:
Updated tests as per [comment here](https://github.com/theforeman/foreman/pull/7105#discussion_r354283920)

jeremylenz commented on webpack/assets/javascripts/react_app/common/IntegrationTestHelper.js:
```diff
@@ -55,15 +55,15 @@ export default class IntegrationTestHelper {
+   * Compare the store with the stored snapshot
+   * @param  {string} description Snapshot description
    */
   takeStoreSnapshot(description) {
```
@sharvit by the way, it says the first argument of `toMatchSnapshot` is a [property matcher object](https://jestjs.io/docs/en/snapshot-testing#property-matchers) and not a string, I'm not sure if `description` is doing anything here?

sharvit replied:
Look like it works, see for example:
- https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/integration.test.js#L46
- https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/__snapshots__/integration.test.js.snap#L48

jeremylenz replied:
@sharvit so it does!  Yay for undocumented API's

sharvit commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModalActions.js:
```diff
@@ -0,0 +1,52 @@
+
+export const addModal = ({ id, open = false }) => (dispatch, getState) => {
+  const { foremanModals } = getState();
+  if (foremanModals && foremanModals[id]) {
```
Can you reuse a selector instead of reading the state directly?

jeremylenz replied:
I can use the selector but I'll still have to pass the state into it.

jeremylenz replied:
Done :white_check_mark:

sharvit commented on webpack/assets/javascripts/react_app/components/ForemanModal/__tests__/ForemanModal.integration.test.js:
```diff
@@ -0,0 +1,96 @@
+    integrationTestHelper.store.dispatch(addModal({ id: 'modal2' }));
+    integrationTestHelper.store.dispatch(addModal({ id: 'modal3' }));
+
+    const modal1 = integrationTestHelper.mount(
```
I'm wondering, instead of manually dispatching `addModal` before mounting. Would it make sense if `ForemanModal` will automatically call it `onMount` and will clean the state `onUnmount`?

jeremylenz replied:
Ah, so have `ForemanModal` itself handle dispatching the `addModal` action for you.  nice idea!

Since ForemanModal is a function component and not a class, I will do it with `useEffect` rather than refactoring to a class component just to use `componentDidMount` and `componentWillUnmount`.

sharvit replied:
Yes! and IMO the `useEffect` should be used inside the `ConnectedForemanModal`

jeremylenz replied:
@sharvit Updated

sharvit commented on webpack/assets/javascripts/react_app/components/ForemanModal/__tests__/ForemanModal.integration.test.js:
```diff
@@ -0,0 +1,96 @@
+
+    // After a Redux action updates the component, this ensures we're looking at
+    // the latest version of the React tree after rerender.
+    const updateWrappers = () =>
```
Using `await IntegrationTestHelper.flushAllPromises();` instead `updateWrappers` would work?

jeremylenz replied:
@sharvit I tried it but the test failed.

sharvit commented on webpack/assets/javascripts/react_app/components/ForemanModal/__tests__/__snapshots__/ForemanModal.integration.test.js.snap:
```diff
@@ -0,0 +1,107 @@
+        "open": true,
+      },
+      "modal2": Object {
+        "open": true,
```
Wondering if it make sense to have 2 modals open at the same time. What would be the result of this scenario in real life? Should we allow it?

jeremylenz replied:
@sharvit It's already in use in https://github.com/Katello/katello/pull/8384 where you can have the ManageManifestModal and the manifest delete confirmation modal open at the same time.  Seems to work fine!

sharvit replied:
Does it work because the css z-index tweaked to allow it in this scenario?
I assume it won't work if I will just open 2 or more modals one by one?

jeremylenz replied:
@sharvit Looking at the elements created by the PF3 component, it seems that all PF3 modals are assigned the same z-index:
```css
.modal {
  z-index: 1050;
}
```
When this is the case, (and assuming they are in the same [stacking context](https://tiffanybbrown.com/2015/09/css-stacking-contexts-wtf/), which I believe they are) the elements will stack in the [same order that they appear in the DOM](https://css-tricks.com/almanac/properties/z/z-index/).  So in this case, since the delete confirmation appears in the markup after the ManageManifestModal, it will be on top.

For this use case (having a main modal on screen and also a confirmation dialog for an action on the main modal) this behavior makes it easy for us, because we don't have to do any manual changes.

For an incorrect use case (trying to have two main modals on the screen) the consumer may see mixed results for which modal ends up on top.  However, since we want to discourage this type of use anyway, I don't think any changes are necessary.

sharvit commented on webpack/assets/javascripts/react_app/components/ForemanModal/ForemanModal.js:
```diff
@@ -1,54 +1,58 @@
-  title: PropTypes.string.isRequired,
+  title: PropTypes.string,
+  id: PropTypes.string.isRequired,
+  openState: PropTypes.shape({
```
Any reason to use `shape`? Why not `open: PropTypes.bool`?

jeremylenz replied:
@sharvit It would be simpler in the component, yes.  But I don't want our Redux state to look like 
```js
myModal: true,
```
instead of 
```js
myModal: { open: true }
```

sharvit replied:
I think it makes sense to do it and use a new selector that will return boolean. `selectIsModalOpen`

jeremylenz replied:
@sharvit Ok, will update

sharvit reviewed (changes requested):
Thank you @jeremylenz , tests looks much better now :+1: 

I have a concern about allowing 2 modals to be open together. I explained more in one of the comments.

jeremylenz commented:
Added Storybook stories and rebased to current develop

xprazak2 commented:
No longer blocked by other PRs I think, removing label.

jeremylenz commented:
Squashed & rebased

sharvit reviewed (dismissed):
Thanks @jeremylenz, looks good and works as expected :1st_place_medal:

mmoll commented:
merged, thanks @jeremylenz!
